File: Python/Array/M-2336. Smallest Number in Infinite Set.py
class SmallestInfiniteSet:
    def __init__(self):
        self.has_nums = [True for _ in range(1000)]

    def popSmallest(self) -> int:
        for i in range(len(self.has_nums)):
            if self.has_nums[i]:
                self.has_nums[i] = False
                return i + 1

    def addBack(self, num: int) -> None:
        self.has_nums[num - 1] = True


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)


obj = SmallestInfiniteSet()
obj.addBack(2)
print(obj.popSmallest())  # 1
print(obj.popSmallest())  # 2
print(obj.popSmallest())  # 3
obj.addBack(1)
print(obj.popSmallest())  # 1
print(obj.popSmallest())  # 4
print(obj.popSmallest())  # 5

File: Array/M-2336. Smallest Number in Infinite Set/M-2336. Smallest Number in Infinite Set.ts
class SmallestInfiniteSet {
  hasNums: boolean[];

  constructor() {
    this.hasNums = new Array(1000).fill(true);
  }

  popSmallest(): number | null {
    for (let i = 0; i < this.hasNums.length; i++) {
      if (this.hasNums[i]) {
        this.hasNums[i] = false;
        return i + 1;
      }
    }
    return null;
  }

  addBack(num: number): void {
    this.hasNums[num - 1] = true;
  }
}

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * var obj = new SmallestInfiniteSet()
 * var param_1 = obj.popSmallest()
 * obj.addBack(num)
 */

var obj = new SmallestInfiniteSet();
obj.addBack(2);
console.log(obj.popSmallest()); // 1
console.log(obj.popSmallest()); // 2
console.log(obj.popSmallest()); // 3
obj.addBack(1);
console.log(obj.popSmallest()); // 1
console.log(obj.popSmallest()); // 4
console.log(obj.popSmallest()); // 5
